### PR TITLE
Enable '.' for nested field in text embedding processor

### PR DESCRIPTION
**Pull request by martin-gaievski**
### Description
Adding support for complex structures in inference processor definition. Main purpose is to improve user experience for cases when object has complex hierarchical structure, so users can apply easier syntax in processor definition.

Example of such new format:

```
"a.b.c.d": "field"
```
or

```
"a.b" : {
   "c.d": "field"
}
```

in this case we will try to look for structure like this in the ingest document:

```
"a" : {
   "b": {
     "c" : {
       "d" : "field"
```

Today we do support only hierarchical type of definition in mapping. It must look exactly like it is in the document:

```
"a" : {
   "b": {
     "c" : {
       "d" : "field"
```  

**Note**: this change affects only source field (left part of the mapping, one that holds value that is a basis for embedding generation). Today logic for the destination field (right part in mapping, field that will store generated embeddings) will be unchanged. As per today's logic that destination field is expected at the same level with the final source field. Example: `"a.b.c: d"` in this case embeddings are inserted in following structure: 
```
"a" : {
   "b" : {
     "d" : [0.1, 0.2 ....]
```

### Issues Resolved
https://github.com/opensearch-project/neural-search/issues/110

### Check List
- [X] New functionality includes testing.
    - [x] All tests pass
- [X] New functionality has been documented.
    - [X] New functionality has javadoc added
- [X] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 1]
```diff
@@ -317,21 +325,45 @@ private void buildMapWithProcessorKeyAndOriginalValueForMapType(
...
             treeRes.put(key, sourceAndMetadataMap.get(parentKey));
         }
     }
 
+    /**
+     * Process the nested key, such as "a.b.c" to "a", "b.c"
+     * @param nestedFieldMapEntry
+     * @return A pair of the original key and the target key
+     */
+    private Pair<String, Object> processNestedKey(final Map.Entry<String, Object> nestedFieldMapEntry) {
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:353]: Do we have a unit test for this?
I don't see any test for this method in `InferenceProcessorTests` file.

Also, this might break backward compatibility for user who is using dot in their key already. Are we fine on that?

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 2]
```diff
@@ -293,20 +298,23 @@ Map<String, Object> buildMapWithTargetKeyAndOriginalValue(IngestDocument ingestD
         return mapWithProcessorKeys;
     }
 
-    private void buildMapWithProcessorKeyAndOriginalValueForMapType(
+    private void buildNestedMap(
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:301]: Would be nice to have unit test for this method

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 3]
```diff
@@ -276,15 +279,17 @@ private void createInferenceListForMapTypeInput(Object sourceValue, List<String>
     }
 
     @VisibleForTesting
-    Map<String, Object> buildMapWithTargetKeyAndOriginalValue(IngestDocument ingestDocument) {
+    Map<String, Object> buildMapWithTargetKeys(IngestDocument ingestDocument) {
         Map<String, Object> sourceAndMetadataMap = ingestDocument.getSourceAndMetadata();
         Map<String, Object> mapWithProcessorKeys = new LinkedHashMap<>();
         for (Map.Entry<String, Object> fieldMapEntry : fieldMap.entrySet()) {
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:285]: I assume field map is what will be defined when user create processor?
Wouldn't it be better to parse and expand the key during creation time rather than ingestion time?

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 4]
```diff
@@ -276,15 +279,17 @@ private void createInferenceListForMapTypeInput(Object sourceValue, List<String>
...
-            String originalKey = fieldMapEntry.getKey();
-            Object targetKey = fieldMapEntry.getValue();
+            Pair<String, Object> processedNestedKey = processNestedKey(fieldMapEntry);
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:286]: Is the fieldMapEntry what user define during processor creation?
If that is the case, can we parse and expand the key during creation time?

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 1 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:353]: let me add the test. For the bwc we should be fine, it's more like an edge case if someone expecting "a.b.c" to be a single field and not the structure. It's one way or another and we're making this change based on requests from users and because this syntax supported in other places in Open search.

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 2 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:301]: ack

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 4 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:286]: yes, it should be possible to construct that map when we create a processor class. let me do the change

Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:286]: Cool. That way, we might be able to reduce some latency during ingestion.

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 1 above)
Comment (heemin32) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:353]: Not about bwc test. Just writing it as unit test friendly would make it easier to maintain.

Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java]: Got it, added unit tests

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 5]
```diff
@@ -62,6 +65,7 @@ public abstract class InferenceProcessor extends AbstractProcessor {
 
     private final Environment environment;
     private final ClusterService clusterService;
+    List<Pair<String, Object>> mappingForNestedFields;
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:68]: Can we make it as private? why should it be package scope?

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 6]
```diff
@@ -84,6 +88,7 @@ public InferenceProcessor(
         this.mlCommonsClientAccessor = clientAccessor;
         this.environment = environment;
         this.clusterService = clusterService;
+        mappingForNestedFields = fieldMap.entrySet().stream().map(this::processNestedKey).collect(Collectors.toList());
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:91]: I didn't mean to just move this in constructor. As we are calling processNestedKey inside execute method anyway, there is no much difference between this code and previous code. Actually, previous code is better...

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 7]
```diff
@@ -317,21 +322,46 @@ private void buildMapWithProcessorKeyAndOriginalValueForMapType(
...
             treeRes.put(key, sourceAndMetadataMap.get(parentKey));
         }
     }
 
+    /**
+     * Process the nested key, such as "a.b.c" to "a", "b.c"
+     * @param nestedFieldMapEntry
+     * @return A pair of the original key and the target key
+     */
+    @VisibleForTesting
```
Comment (heemin32) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:350]: I am fine for this PR but @VisibleForTest is not the best way for unit test because it makes it hard to write unit test for `execute` method as there is no easy way to mock the sub methods.

The ideal way should be something like
```
public InferenceProcessor (MyClass myclass) {
  this.myclass = myclass;
}

public void execute() {
  myclass.doSomething();
  myclass.doAnother();
  ...
}
```

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 6 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java]: got it, yeah there is no easy way to remove that call from recursion. reverting back to the previous state

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 5 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java]: not relevant as I'm reverting back to previous approach

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 7 above)
Comment (martin-gaievski) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:349]: agree, we can do such refactoring for tests specifically where we do complex logic that needs matrix of all permutations of params and/or input.

**Review by heemin32 — APPROVED** [review]

**Review by vibrantvarun — APPROVED** [review]
Looks good. Curious how nested field is processed currently without this change?

**Comment by martin-gaievski** [comment]
> Looks good. Curious how nested field is processed currently without this change?

Today we treat field name separated with `.` as a single field name, if someone put `"a.b" : "c"` we will pickup the field only if doc has field `"a.b"`, not `"a": { "b"}`. There is a workaround, user can put mapping as a structured json, I put example in PR description. In this case UX isn't great for cases when objects have complex structure.
